### ISSUE TITLE
`RET504` not triggered with arbitrary calls

### ISSUE THREAD

**Issue by dhruvmanila** [astral-sh/ruff#4242]
What about the following code:

```python
def func() -> int:
    match True:
        case True:
            x = 5
            return x
        case _:
            raise ValueError("Error")
```

I don't get `RET504` here as well. Is that the expected behavior?

_Originally posted by @Apakottur in https://github.com/charliermarsh/ruff/issues/4240#issuecomment-1536067821_

**Comment by dhruvmanila** [comment]
After giving a quick look, it's occurring as the `Visitor` is not going through the `match` statement.

**Comment by dhruvmanila** [comment]
Scratch my initial analysis, visitor pattern means that it'll travel to every node _unless_ explicitly avoided. We're going through the match statement, but the problem is two fold:
1. `match` is a mutually exclusive statement. This means that a binding in one case can't be referenced in another case. Visitor is not scope aware.
2. The comment below means that we're considering `x` to be referenced inside the `ValueError` call which is what is creating this false negative.

https://github.com/charliermarsh/ruff/blob/bb2cbf1f2549c7472f0caf913ea4896b2aae0dc4/crates/ruff/src/rules/flake8_return/visitor.rs#L172-L183

Solving (2) seems to be a bit difficult but I think for the short term we can avoid increasing the reference count for builtin calls.

For the long term, the scope will have to be implemented which can be used to increase the reference count of variables which the function has access to, either directly through the arguments or from an outer scope. We could even go further and traverse the function body to check if it's actually being referenced but there might be a challenge in differentiating between referencing the same variable name defined inside the function vs actually referencing it from outside the scope.

These are just some of my thoughts after debugging this issue. I think this might not be a _good first issue_.

**Comment by charliermarsh** [comment]
I am tempted to change this rule such that it only detects:

```
x = 5
return x
```

That is: such that it only detects "assignment followed by return".
